RFC/src/RFC-0250_Covenants.md

```diff
@@ -0,0 +1,468 @@
+    filter_relative_height(10),
+)
+```
+
+is represented in hex bytes as `23 30 01 a8b3f48e39449e89f7ff699b3eb2b080a2479b09a600a19d8ba48d765fe5d47d 35 07 0a`.
+Let's unpack that as follows:
+```
+23 // xor - consume two covenant args
+30 // filter_output_hash_eq - consume a hash arg
+01 // 32-byte hash
+a8b3f48e39449e89f7ff699b3eb2b080a2479b09a600a19d8ba48d765fe5d47d // data
+// end filter_output_hash_eq
+35 // 2nd covenant - filter_relative_height
+07 // varint
+0A // 10 
+// end varint, filter_relative_height, xor
+```
+
+Some functions can take any number of arguments, such as `filter_fields_hashed_eq` which defines the `Fields` type.
+This type is encoded first by its byte code `34` followed by a varint encoded number that indicates the number
+of field identifiers to consume. To mitigate misuse, the maximum allowed arguments are limited.
+
+### Covenant Validation
+
+A covenant and therefore the block/transaction MUST be regarded as invalid if:
+
+1. an unrecognised byte code is encountered
+2. the end of the byte stream is reached unexpectedly
+3. there are bytes remaining on the stream after interpreting
+4. an invalid argument type is encountered
+5. the `Fields` type encounters more than 9 arguments (i.e. the number of fields tags available)
+6. the depth of the calls exceeds 16. 
+
+### Consensus changes
+
+The covenant is executed once all other validations, including [TariScript], are complete. This ensures that
+invalid transactions in a block cannot influence the results. 
+
+## Considerations 
+
+### Complexity
+
+This introduces additional validation complexity. We avoid stacks, loop, and conditionals (covenants are basically 
+one conditional), there are overheads both in terms of complexity and performance as a trade-off for the 
+power given by covenants. 
+
+The worst case complexity for covenant validation is `O(num_inputs*num_outputs)`, although as mentioned above
+validation for each input can be executed in parallel. To compensate for the additional workload the network 
+encounters, use of covenants should incur heavily-weighted fees to discourage needlessly using them.
+
+### Cut-through 
+
+The same arguments made in the [TariScript RFC](./RFC-0201_TariScript.md#cut-through) for the need to prevent
+[cut-through] apply to covenants.
+
+### Chain analysis
+
+The same arguments made in the [TariScript RFC](./RFC-0201_TariScript.md#fodder-for-chain-analysis) apply.
+
+### Security
+
+As all outputs in a block are in the scope of an input to be checked, any unrelated/malicious output in a block 
+_could_ pass an unrelated covenant rule if given the chance. A secure covenant is one that _uniquely_ identifies 
+one or more outputs.
+
+## Examples
+
+### Now or never
+
+Spend within 10 blocks or burn
+
+```
+not(filter_relative_height(10))
+```
+
+Note, this covenant may be valid when submitted to the mempool, but invalid by the time it is put in a block for
+the miner.
+
```

Comment:
This is a good example of something you might want to do with this, but this is something I would argue we should rather do with TariScript. 

The difference being:
TariScript: Controlling when you may spend your already mined UTXO
Convenants: Controlling to what you may spend your UTXO. 

And here they focus on the rules relating to the input and output respectively and I think we should stick to these as far as possible to simplify things.

Reply:
Yeah, I'm 50/50 on needing that covenant because it means we have to revalidate before assembling the block. This isn't the best example. The positive 50% is thinking that it seems useful to be able to either run covenant `A` or `B` depending on maturity. This allows spending to outputs a certain way before 100 blocks and another way after.

Maybe something like a vault which allows you access to funds while locked in the vault (say for 100 blocks) but ensures that the covenant protects the funds for those 100 blocks (by allowing you to burn the funds if an attacker pwned your private keys, discouraging the attacker because you can ensure that no one gets the funds). I didn't manage to come up with an example that covenant but I think it's possible.